lizard_auth_client/tests.py

```diff
@@ -1071,6 +1072,114 @@ def test_organisation_index_view_superuser(self):
+                    args=[self.organisation_1.id]))
+        request.session = {}
+        request.user = self.user_1
+        request.user.is_superuser = True
+        request.POST = post_data
+
+        return views.ManageUserAddView.as_view()(
+            request, self.organisation_1.id)
+
+    def test_organisation_add_totally_new_user(self):
+        """A manager should be able to add an user
+           with an unknown email address and username
+        """
+        with self.get_new_user_sso_mock():
+            response = self.get_new_user_response(
+                {'username': 'root_new',
+                 'first_name': 'willie',
+                 'email': 'noreply_new@example.com'})
+
+            # Should return an HTTP redirect to detail page
+            self.assertEqual(response.status_code, 302)
+
+    def test_organisation_add_known_exact_match_user(self):
+        """A manager should be able to add an user
+           with an exactly matching email-address/username combination
+        """
+        with self.get_new_user_sso_mock() as sso_mock:
+            # Inject error raised by _new_user_sso_post
+            sso_mock.side_effect = views.UserNotCreatedError(
+                "User could not be created",
+                response=mock.Mock(
+                    status_code=200,
+                    __getitem__=lambda x, y: {
+                        'username': 'root',
+                        'email': 'noreply@example.com'}))
+
+            response = self.get_new_user_response(
+                {'username': 'root',
+                 'first_name': 'willie',
+                 'email': 'noreply@example.com'})
+
+            # Should return an HTTP redirect to detail page
+            self.assertEqual(response.status_code, 302)
+
+    def test_organisation_add_user_with_existing_email_address(self):
+        """A manager should not be able to add an user with an existing
+           email address
+        """
+        with self.get_new_user_sso_mock() as sso_mock:
+            sso_mock.side_effect = views.UserNotCreatedError(
+                "User could not be created",
+                response=mock.Mock(
+                    status_code=200,
+                    __getitem__=lambda x, y: {
+                        'username': 'root_existing',
+                        'email': 'noreply@example.com'}))
+
+            response = self.get_new_user_response(
+                {'username': 'root_new',
+                 'first_name': 'willie',
+                 'email': 'noreply@example.com'})
+
+            # Should return form validation error indicating
+            # that a user with "root_existing" username already
+            # exists for the given email-address
+            self.assertEqual(response.status_code, 200)
+            self.assertEqual(
+                response.context_data['form'].errors['username'],
+                ['The given username does not match the email address, '
```

Comment:
Dit ziet eruit alsof je op één of andere manier toch de gebruikersnaam terugkrijgt die bij het email adres "hoort".
Dat was toch niet de bedoeling? Dan zou lizard-auth-server toch een user object teruggeven i.p.v. alleen een 409 error.

Workflow zou moeten zijn dat je EERST op emailadres zoekt en DAARNA pas een user object aan probeert te maken. Zoiets. Dus het max dat je kan zeggen is "username bestaat al, kies een iets andere".

Reply:
Ipv. te zoeken probeert hij direct de user aan te maken op de SSO server met emailadres/username combi, het volgende kan dan gebeuren:

1) emailadres en username bestaan niet op SSO server -> alles ok, maak nieuwe user aan op SSO en lokaal 
2) emailadres bestaat al met zelfde username op SSO server -> alles ok, maak lokaal aan
3) emailadres bestaat al met andere username op SSO server -> Niet ok, laat validatie error zien dat er andere usename wordt verwacht
4) emailadres bestaat niet maar username wel op SSO server -> Niet ok, laat validatie error zien dat er username al in gebruik is

In de "test_organisation_add_user_with_existing_email_address" wordt een error geraised op status 200, we willen namelijk een 201 (Created) terugkrijgen. 200 betekent dat het e-mail adres al in gebruik is (scenario nummer 3), hij geeft dan een validatie error met welke username wordt verwacht voor dat e-mail adres. 

De 409 wordt alleen teruggegeven als **alleen** de username al in gebruik is, als het email adres gelijk is dan krijg je altijd een 200 met user object in JSON ongeacht of username al in gebruik is ja/nee.